### PR TITLE
[Dictionary] Update BPE

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
BPE Helpers in ParlAI were disparate, cluttering core, and just a tad bit messy.

I've done some work to compile all the BPE helpers in one file, abstracting what is necessary from each, and making the API calls in `dict.py` (in my opinion) nicer.

As a result, I have been able to implement a "GPT2 Standin" for the HuggingFace BPE Tokenizer, for systems where one does not have access to HuggingFace's `tokenizers` library (e.g. in [REDACTED] where we need to run models for [REDACTED])

**Testing steps**
Ran existing `test_dict.py` tests, as well as added one for my own sake.

I also manually tried loading a model in interactive with the GPT2 Standin that was trained with HF's bpe tokenizer 

**Logs**
```
(env with tokenizers) $ python examples/interactive.py <args>...

Enter Your Message: how's it going i love my dog
/private/home/kshuster/MyParlAI/parlai/core/torch_ranker_agent.py:718: UserWarning: [ Executing eval mode with a common set of fixed candidates (n = 223649). ]
  "(n = {}). ]".format(mode, len(self.fixed_candidates))
[ImagePolyencoder]: i love dogs what kind do you have
Enter Your Message: a golden retriever! how bout you?
[ImagePolyencoder]: me too, a golden retriever
```

```
(env WITHOUT tokenizers) $ python examples/interactive.py <args>...
.
.
.
/private/home/kshuster/MyParlAI/parlai/utils/bpe.py:62: UserWarning:

--------------------------------------------------

WARNING: You have chosen to use Huggingface's tokenizer.
Please install HuggingFace tokenizer with: pip install tokenizers.
For now, defaulting to the GPT2Tokenizer.

--------------------------------------------------


  ''
.
.
.
Enter Your Message: how's it going i love my dog
/private/home/kshuster/MyParlAI/parlai/core/torch_ranker_agent.py:718: UserWarning: [ Executing eval mode with a common set of fixed candidates (n = 223649). ]
  "(n = {}). ]".format(mode, len(self.fixed_candidates))
[ImagePolyencoder]: i love dogs what kind do you have
Enter Your Message: a golden retriever! how bout you?
[ImagePolyencoder]: me too, a golden retriever
```

**Other information**
I am genuinely curious how the GPT2 tokenizer handles dict special tokens, as I am not sure I quite understand that...

### CODE REVIEW

File: tests/test_dict.py [hunk 1]
```diff
@@ -394,3 +400,107 @@ def test_build_split(self):
 
     def test_build_bpe(self):
         self._run_test({'dict_tokenizer': 'bpe', 'max_tokens': 50})
+
+
+class TestGpt2HFInteropt(unittest.TestCase):
```
Comment (emilydinan) [inline tests/test_dict.py:405]: nit: interop

**Comment by klshuster** [comment]
@emilydinan That is a great question - who, indeed, is using the gpt2 helper? @stephenroller just recently informed me that a few of his older models were using this helper; in general, this is what is used if you specify the `--dict-tokenizer` to be gpt2 (i.e. any model that wants to use our regular `DictionaryAgent` with a gpt2 dictionary/tokenizer). The dictionary you linked is used _specifically_ by our gpt2 model. It also relies on the HF `tokenizers` library

**Comment by stephenroller** [comment]
(maybe run long tests on your devfair)

**Comment by stephenroller** [comment]
I think we could think of a better name than gpt2standin. it really is more like SlowBytelevelBPE

File: parlai/core/dict.py [hunk 2]
```diff
@@ -624,22 +588,18 @@ def save(self, filename=None, append=False, sort=True):
...
+                self.bpe.copy_codecs_file(filename + '.codecs')
             if sort:
                 self.sort(trim=False)
-        elif self.tokenizer == 'gpt2':
-            # never remove or sort tokens from gpt2
-            pass
-        elif self.tokenizer == 'bytelevelbpe':
-            # never remove or sort tokens from bytelevelbpe, it should be the
-            # same as the hugging face tokenizer
+        elif self.tokenizer in ['gpt2', 'bytelevelbpe', 'gpt2_standin']:
```
[the remark this answers is not in the document]
Comment (klshuster) [reply parlai/core/dict.py:601]: sure i can look into making this better

**Comment by klshuster** [comment]
Yeah, `GPT2Standin` was a stand in (😉) name anyway, I like `SlowBytelevelBPE`

**Comment by klshuster** [comment]
Will run long tests on devfair

**Comment by klshuster** [comment]
All relevant long tests pass (barring some ones that do not, fixes incoming in separate PR as they are unrelated to this PR)

**Pull request merged**